### PR TITLE
ome-training-2: Replacement for current outreach

### PR DESCRIPTION
**Pull request by manics**
Main changes:
- OMERO.server runs from a virtualenv
- OMERO 5.5.1
- PostgreSQL 10.8
- Includes a convenience wrapper `/usr/local/bin/omero`

Requires:
- https://github.com/ome/ansible-role-omero-server/pull/36
- https://github.com/ome/ansible-role-omero-web/pull/21
----
## Updated 2019-09-24
Following discussion with @pwalczysko this is the plan:
- From now until 27 October 2019 `outreach` and `workshop` will be treated as independent servers.
- On or after 28 October 2019 the existing `outreach` will be taken out of service, to be reimaged and rebuilt to be identical to the new `workshop`. At this point the `omero/training-server` directory will be deleted, `omero/training-server-2` directory will be renamed `omero/training-server`, and both servers (`outreach`, `workshop`) will be managed as a single group `ome-outreach`
- In the meantime changes can be made independently to `outreach` as required by @pwalczysko. If necessary these changes will be ported to `workshop`.

I've rebased and updated this and the related private inventory PR.
This ensures we can faithfully represent the state of all production systems at all times.

To view changes:
- `diff -Nur omero/training-server omero/training-server-2`
- `diff -Nur molecule/omero-training-server molecule/omero-training-server-2`
----
## Updated 2019-10-31
Old `ome-training-outreach` server has been replaced by `ome-training-1`, outreach.openmicroscopy.org still works.

### CODE REVIEW

**Comment by manics** [comment]
OMERO was restored as follows:
```bash
systemctl stop omero-server
sudo -u postgres pg_restore --clean -d omero omero-database-omero-20190710-154903-139204.pgdump
cd /
sudo tar -xf ~/outreach-omero-20190710/outreach-OMERO-20190710-1550.tar
systemctl start omero-server
```

**Comment by joshmoore** [comment]
The wrapper may need to overwrite the original location due to hard-coded scripts using `/opt/omero/server/OMERO.server/bin/omero`

**Comment by kennethgillen** [comment]
Interesting [drop of the fqdn hostname](https://github.com/ome/prod-playbooks/pull/186/files#diff-a97d1f90899e5c57cc9250a49bf2d4efR3). Deployed to a different infrastructure?

**Comment by joshmoore** [comment]
I'd assume it's just a group that's in a different PR.

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [TRAINING-merge#333](https://ci.openmicroscopy.org/job/TRAINING-merge/333/). See the [console output](https://ci.openmicroscopy.org/job/TRAINING-merge/333/consoleText) for more details.
Possible conflicts:
  - Upstream changes
    - requirements.yml
    - omero/training-server/playbook.yml

--conflicts

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [TRAINING-merge#334](https://ci.openmicroscopy.org/job/TRAINING-merge/334/). See the [console output](https://ci.openmicroscopy.org/job/TRAINING-merge/334/consoleText) for more details.
Possible conflicts:
  - Upstream changes
    - requirements.yml
    - omero/training-server/playbook.yml

--conflicts

**Comment by manics** [comment]
This has been deployed to `ome-training-2.openmicroscopy.org` (also known as `workshop.openmicroscopy.org`), but will need rebasing after https://github.com/ome/prod-playbooks/pull/206

**Comment by manics** [comment]
Note updated description

**Comment by manics** [comment]
Rebased following merging of https://github.com/ome/prod-playbooks/pull/206

`diff -Nur omero/training-server/playbook.yml omero/training-server-2/playbook.yml`
```diff
--- omero/training-server/playbook.yml	2019-09-26 10:11:33.000000000 +0100
+++ omero/training-server-2/playbook.yml	2019-09-26 10:38:19.000000000 +0100
@@ -1,6 +1,9 @@
-# Install OMERO.server and OMERO.web on localhost
+---
+# The OME training server(s)
+# Set https_letsencrypt_enabled to True in production to ensure certbot
+# handles the letsencrypt certificate setup
 
-- hosts: outreach.openmicroscopy.org
+- hosts: ome-outreach
   pre_tasks:
 
     - name: Install Make Movie script Prerequisite | MEncoder - Repo
@@ -41,13 +44,18 @@
       - user: omero
         password: omero
         databases: [omero]
-      postgresql_version: "9.6"
+      postgresql_version: "10"
 
     - role: ome.versioncontrol_utils
 
     - role: ome.nfs_mount
 
     - role: ome.omero_server
+      omero_server_virtualenv: yes
+      omero_server_python_addons:
+        - "omero-cli-duplicate=={{ omero_cli_duplicate_release }}"
+        - "omero-cli-render=={{ omero_cli_render_release }}"
+        - "omero-metadata=={{ omero_metadata_release }}"
 
     - role: ome.omero_web
       # omero_web_config_set:
@@ -70,7 +78,7 @@
         state: directory
 
     # Inspired by
-    # https://github.com/openmicroscopy/prod-playbooks/blob/7695bf8b1c4f0ae4314524366b91c300c1fdb30b/nightshade-webclients.yml#L119
+    # https://github.com/ome/prod-playbooks/blob/7695bf8b1c4f0ae4314524366b91c300c1fdb30b/nightshade-webclients.yml#L119
     - name: omero-web nginx nested include
       become: yes
       lineinfile:
@@ -105,15 +113,6 @@
         name: docker-python
         state: present
 
-    - name: Omero.cli plugins | plugin install via pip & pypi
-      become: yes
-      pip:
-        name:
-        - "omero-cli-duplicate=={{ omero_cli_duplicate_release }}"
-        - "omero-cli-render=={{ omero_cli_render_release }}"
-        - "omero-metadata=={{ omero_metadata_release }}"
-        state: present
-
     - name: Create a figure scripts directory
       become: yes
       file:
@@ -251,6 +250,15 @@
         insertbefore: BOF
         line: "+:{{ omero_server_system_managedrepo_group }}:ALL"
 
+    - name: bin/omero wrapper
+      become: yes
+      copy:
+        src: files/bin-omero
+        dest: /usr/local/bin/omero
+        owner: "root"
+        group: "root"
+        mode: 0755
+
     - name: Run docker for ldap
       become: yes
       docker_container:
@@ -338,4 +346,3 @@
     external_nic: "{{ ansible_default_ipv4.interface }}"
 
 - include: letsencrypt.yml
-  when: not (molecule_test | default(False))
```

**Comment by joshmoore** [comment]
> From now until 27 October 2019 outreach and workshop will be treated as independent servers.

This explains the cut-n-paste?  If so, is there likely to be so much divergence between the two?

**Comment by manics** [comment]
The plan now is to re-image outreach so it will be identical to workshop as defined by this PR

**Comment by manics** [comment]
Rebased to include https://github.com/ome/prod-playbooks/pull/211

File: requirements.yml [hunk 1]
```diff
@@ -52,8 +52,11 @@
 - src: ome.omero_python_deps
   version: 2.0.1
 
+- src: openmicroscopy.omero-user
```
Comment (sbesson) [inline requirements.yml]: Why is this necessary?

File: molecule/omero-training-server-2/molecule.yml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+driver:
+  name: docker
+lint:
+  name: yamllint
+  # TODO: enable
+  enabled: False
+platforms:
+  - name: ome-outreach
+    image: centos/systemd
+    image_version: latest
```
Comment (sbesson) [inline molecule/omero-training-server-2/molecule.yml]:
```suggestion
```

File: requirements.yml (hunk 1 above)
Comment (manics) [reply requirements.yml]: Probably a rebasing error

**Comment by pwalczysko** [comment]
Which playbook should I run, the one under training-server or the one under training-server-2 ?

**Comment by manics** [comment]
At the moment neither, I'm in the middle of updating outreach. Once that's done they'll only be one playbook.

**Comment by pwalczysko** [comment]
@sbesson it seems that my testing is finished for now here, see https://openmicroscopy.slack.com/archives/C0K5ME1EW/p1572522697009600

**Comment by pwalczysko** [comment]
I am afraid this is too unclear for me to try to run something against the training servers - these are production. Happy to be told otherwise, as well as please tell me what playbook to run and how to point it to which servers (like outreach, workshop, or both) or is that taken care of already ? @sbesson @manics

**Comment by pwalczysko** [comment]
Thinking about it further, I guess  it would be good to merge this. I will not be running the original playbook here against workshop server. I need to upgrade the apps - hence, need to open a PR for that - only then run the playbook. I guess outreach server could serve as.a testing server, and then, if all is good, we can re-run the playbook against both (outreach and workshop).
cc @manics @sbesson ?

**Comment by sbesson** [comment]
I think the strategy proposed in https://github.com/ome/prod-playbooks/pull/186#issuecomment-548748004 looks sane. As far as I can tell, the playbook has successfully been run against the production training server so there is no reason to keep the old playbook.

Leaving @pwalczysko to perform follow-up tests upgrading the deployment as necessary in preparation of the upcoming workshop.
